### PR TITLE
[SYCL] Cache sub-devices early such they can be seen by interop API

### PR DESCRIPTION
**Pull request by smaslov-intel**
Signed-off-by: Sergey V Maslov <sergey.v.maslov@intel.com>

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -1986,66 +2015,27 @@ pi_result piDevicePartition(pi_device Device,
...
-    }
-    if (OutNumDevices)
-      *OutNumDevices = NumDevices;
-    return PI_SUCCESS;
+  // Devices cache is normally created in piDevicesGet but still make
+  // sure that cache is populated.
+  //
+  pi_result Res = populateDeviceCacheIfNeeded(Device->Platform);
+  if (Res != PI_SUCCESS) {
+    return Res;
```
Comment (bso-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:2023]: Does this means that you always partition the devices into the maximum number of sub-devices, ignoring the user's requested number of devices?

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:2023]: Yes, I have actually added assert that checks this. See line 2031.

Comment (bso-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:2023]: Oh. I see. So, initially you partition fully and keep them in the cache.
Next time if the user requests less than the maximum, we can return only the desired number of sub-devices.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -1986,66 +2015,27 @@ pi_result piDevicePartition(pi_device Device,
...
   }
 
-  // Get the number of subdevices available.
-  // TODO: maybe add interface to create the specified # of subdevices.
-  uint32_t Count = 0;
-  ZE_CALL(zeDeviceGetSubDevices, (Device->ZeDevice, &Count, nullptr));
-
   if (OutNumDevices) {
-    *OutNumDevices = Count;
+    *OutNumDevices = Device->SubDevices.size();
```
Comment (bso-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:2027]: It looks to me here that we always return all maximum number of sub-devices regardless of user's request.
Is this right?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -1986,66 +2015,27 @@ pi_result piDevicePartition(pi_device Device,
...
   }
 
-  if (!OutDevices) {
-    // If we are not given the buffer, we are done.
-    return PI_SUCCESS;
-  }
+  if (OutDevices) {
+    // TODO: Consider support for partitioning to <= total sub-devices.
+    PI_ASSERT(NumDevices == Device->SubDevices.size(), PI_INVALID_VALUE);
```
Comment (bso-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:2032]: I think we don't need this assert here.
We can just return the requested number of sub-devices (NumDevices).
Please note that the order of sub-device handles returned from L0 driver is always the same order.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 4]
```diff
@@ -1986,66 +2015,27 @@ pi_result piDevicePartition(pi_device Device,
...
-        }
-        OutDevices[I] = PiSubDevice.get();
-        // save pointers to sub-devices for quick retrieval in the future.
-        Device->SubDevices.push_back(PiSubDevice.get());
-        Platform->PiDevicesCache.push_back(std::move(PiSubDevice));
-      }
+    for (uint32_t I = 0; I < NumDevices; I++) {
+      OutDevices[I] = Device->SubDevices[I];
+      // reusing the same pi_device needs to increment the reference count
+      PI_CALL(piDeviceRetain(OutDevices[I]));
```
Comment (bso-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:2037]: Here we are returning the requested number of devices correctly.
So, I guess the assert above at line 2032 is unnecessary.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:2032]: Today, the only supported partitioning doesn't have a way tp specify # of devices. So truly, SYCL RT will always ask for all sub-devices:
``` C++
        auto SubDevicesDomainNuma = rootdev.create_sub_devices<
            sycl::info::partition_property::partition_by_affinity_domain>(
            sycl::info::partition_affinity_domain::numa);
```

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:2032]: I just wanted to check this assumption explicitly.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 5]
```diff
@@ -1350,7 +1352,6 @@ pi_result piDevicesGet(pi_platform Platform, pi_device_type DeviceType,
 
   PI_ASSERT(Platform, PI_INVALID_PLATFORM);
 
-  std::lock_guard<std::mutex> Lock(Platform->PiDevicesCacheMutex);
```
Comment (bso-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1353]: Don't we need PiDevicesCacheMutex lock at line 1362?

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (bso-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:2032]: I thought we support partitioning of sub-devices, which means partitioning multiple levels.
In Sycl spec 2020 page 64,
<pre>
A SYCL device can be partitioned into multiple SYCL devices, by calling the create_sub_devices() member
function template. The resulting SYCL devices are considered sub devices, and it is valid to partition
these sub devices further.
</pre>

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 5 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1353]: I don't think we need to lock the map for read-only access

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:2032]: One should first query the supported partitioning, and Level-Zero would only return the one I showed.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 5 above)
Comment (bso-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1353]: It is going to be a problem if there is another thread writing to the cache while this thread is reading the cache at the same time.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (bso-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:2032]: Do you mean partition_by_affinity does not support partitioning multiple levels?
If so, could you add a comment about this fact for other people?
"We currently support only partition_by_affinity which does not allow partitioning multiple levels. So we always return the maximum number of sub-devices."
That will be very useful to understand the logic behind this code.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 5 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1353]: No one should be reading the cache until it is populated. This is guaranteed by calling to populateDeviceCacheIfNeeded before attempting the read

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:2032]: what is "multiple levels" in your statement?

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 5 above)
Comment (bso-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1353]: Hmm.. I thought piextDeviceCreateWithNativeHandle() can call populateDeviceCacheIfNeeded() in one thread.
And another thread can call piDevicesGet() at the same time.
Is this impossible?

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1353]: It possible and would be OK. Both piextDeviceCreateWithNativeHandle and piDevicesGet first call to populateDeviceCacheIfNeeded, which does the lock. So only one thread will populate the cache and others will wait. After that all accesses to the cache are read-only

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (bso-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:2032]: What I mean by multiple levels is calling partition_device() on the sub-devices which are already partitioned before.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 5 above)
Comment (bso-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1353]: I see. It makes sense.
piDevicesGet() will wait until the other writer is done.
Thanks for your clarification.

**Review by bso-intel — DISMISSED** [review]
LGTM.
Nit: it would be great if you add some comments about why we always return max number of sub-devices regardless of the user's request.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:2032]: > partition these sub devices further.

That's not [yet] supported. I will add a comment about why we always expect all devices requested.

**Comment by smaslov-intel** [comment]
@bso-intel : I've additionally removed the lock from getDeviceFromNativeHandle. please review

**Review by bso-intel — APPROVED** [review]
LGTM. Thanks.

**Review by bso-intel — APPROVED** [review]
LGTM

**Pull request merged**